piqe_ocp_lib/api/resources/ocp_cluster_versions.py

```diff
@@ -39,37 +44,163 @@ def update_cluster_version(self, cv_body):
+        except IndexError:
+            logger.debug("Cluster already at latest version.")
+
+    def update_channel(self, channel: str):
+        """
+        Change update channel
+        :param channel: (str) name of the channel
+        :return: (ClusterVersion) object
+        """
+        cv_body = {"spec": {"channel": channel}}
+        cv_body = self._build_spec(cv_body)
+        return self.update_cluster_version(cv_body)
+
+    def upgrade_cluster_version(self, version: str, force: bool = False, timeout: int = 0) -> Dict:
+        """
+        Upgrade cluster to desired version
+        :param force: whether to force the update or not
+        :param timeout: minutes to wait for upgrade completion
+        :return: ClusterVersion API response
+        """
+        def wait_until_upgraded(timeout: int):
+            """
+            Polling of CVO status until upgrade is completed or fail otherwise
+            :return: CVO object OR None
+            """
+            ends_at = datetime.now() + timedelta(minutes=timeout)
+            def check_upgrade(cluster_version):
+                """
+                Helper function to search for correct entry and check it's state
+                May be replaced by a JSON query language.
+                """
+                search_entry = None
+                for history in cluster_version.status.history:
+                    if history.version == version:
+                        search_entry = history
+                        break
+                return search_entry.state == "Completed"
+
+            while datetime.now() < ends_at:
+                cluster_version = self.get_cluster_version()
+
+                if check_upgrade(cluster_version):
+                    logger.info("Cluster upgraded successfully.")
+                    return cluster_version
+                else:
+                    sorting_key = lambda x: datetime.strptime(x.lastTransitionTime, "%Y-%m-%dT%H:%M:%SZ")
+                    latest_status = sorted(cluster_version.status.conditions, key=sorting_key, reverse=True)[0]
+                    logger.info(f"Upgrade in progress, checking again in {CLUSTER_POLLING_SECONDS_INTERVAL} seconds.")
+                    logger.info(f"The latest event is {latest_status.type} - {getattr(latest_status, 'message', '')}")
+                    time.sleep(CLUSTER_POLLING_SECONDS_INTERVAL)
+            else:
+                # Polling exhausted
+                logger.error(f"Failed to verify cluster upgrade during {timeout} minutes.")
+
+        cv_body = self._build_spec({
+            "spec": {
+                "desiredUpdate": {
+                    "force": force,
+                    "version": version,
+                }
+            }
+        })
+
+        ocp_cv_upgrade_response = self.update_cluster_version(cv_body)
 
-        return cluster_id
+        return wait_until_upgraded(timeout) if timeout else ocp_cv_upgrade_response
```

Comment:
Similar, separation of logic for readability.

Reply:
I think this is okay for now, as it is working. We can make any agreed upon revisions in a future PR.